File: java/src/merge_sort.java
import java.util.*;

public class merge_sort {
    public static int[] Merge(int[] left, int[] right) {
        int l = 0;
        int r = 0;
        int[] merge = new int[left.length + right.length];
        while (l < left.length && r < right.length) {
            if (left[l] < right[r]) {
                merge[l + r] = left[l];
                l++;
            } else {
                merge[l + r] = right[r];
                r++;
            }
        }
        if (l < left.length)
            System.arraycopy(left, l, merge, l + r, left.length - l);
        else
            System.arraycopy(right, r, merge, l + r, right.length - r);

        return merge;
    }

    public static int[] MergeSort(int[] list) {
        if (list.length <= 1)
            return list;

        else {
            int mid = (int) (list.length / 2);
            int[] left = Arrays.copyOf(list, mid);
            int[] right = Arrays.copyOfRange(list, mid, list.length);
            int[] new_left = MergeSort(left);
            int[] new_right = MergeSort(right);

            return Merge(new_left, new_right);
        }
    }

    public static void main(String[] args) {
        int[] scores = { 5, 1, 0, 2, 7, 3 };
        System.out.println(Arrays.toString(MergeSort(scores)));
    }
}

File: c/merge_sort.c
#include <stdio.h>
#include <string.h>
#include "sort_test.c"

int *array_copy(int dest[], int src[], int dest_start, int count)
{
    for (int i = 0; i < count; i++)
        dest[dest_start + i] = src[i];

    // print_array(dest, dest_start + count);
    // printf("\n");
    return dest;
}

int *merge(int left_part[], int right_part[], int left_size, int right_size)
{
    int l = 0;
    int r = 0;
    int left[left_size];
    int right[right_size];
    memcpy(left, left_part, left_size * 4);
    memcpy(right, right_part, right_size * 4);

    int merge_list[left_size + right_size];

    while (l < left_size && r < right_size)
    {
        if (left[l] < right[r])
        {
            merge_list[l + r] = left[l];
            l++;
        }
        else
        {
            merge_list[l + r] = right[r];
            r++;
        }
    }
    /*printf("\n");
    print_array(merge_list, l + r);
    printf("\n");
    print_array(left, left_size);
    printf("\n");
    print_array(right, right_size);
    printf("\n");*/

    if (l < left_size)
        // return array_copy(merge_list, left + l, l + r, left_size - l);
        memcpy(merge_list + l + r, left + l, (left_size - l) * 4);
    else
        // return array_copy(merge_list, right + r, l + r, right_size - r);
        memcpy(merge_list + l + r, right + r, (right_size - r) * 4);

    return merge_list;
}

int *merge_sort(int list[], int size)
{
    if (size < 2)
        return list;

    int mid = (int)size / 2;
    int left[mid];
    int right[size - mid];
    memcpy(left, list, mid * 4);
    memcpy(right, list + mid, (size - mid) * 4);
    /*printf("split starts\n");
    print_array(left, mid);
    printf("\n");
    print_array(right, size - mid);
    printf("\n");
    printf("split ends\n");*/

    int *new_left = merge_sort(left, mid);
    int *new_right = merge_sort(right, size - mid);
    /*printf("merge starts\n");
    print_array(new_left, mid);
    printf("\n");
    print_array(new_right, size - mid);
    printf("\n");
    printf("merge ends\n");*/

    return merge(new_left, new_right, mid, size - mid);
}

void main()
{
    // sort_test(merge_sort);
    int sort_list[] = {6, 1, 2, 0, 5};
    int *index = merge_sort(sort_list, sizeof(sort_list) / 4);

    print_array(index, 5);
}